### PR TITLE
Implemented cancellation support

### PR DESCRIPTION
**Pull request by Artur-Laskowski**
User is now able to click cancel button to stop processing of current operation.

### CODE REVIEW

File: src/ReferenceAnalyzer.Core/IReferenceAnalyzer.cs [hunk 1]
```diff
@@ -14,5 +15,11 @@ public interface IReferenceAnalyzer
         Task<IEnumerable<string>> Load(string solution);
         IAsyncEnumerable<ReferencesReport> AnalyzeAll();
         IAsyncEnumerable<ReferencesReport> Analyze(IEnumerable<string> projects);
+
+        IAsyncEnumerable<ReferencesReport> AnalyzeAll(string solutionPath, CancellationToken token);
+        Task<ReferencesReport> Analyze(string target, CancellationToken token);
+        Task<IEnumerable<string>> Load(string solution, CancellationToken token);
+        IAsyncEnumerable<ReferencesReport> AnalyzeAll(CancellationToken token);
+        IAsyncEnumerable<ReferencesReport> Analyze(IEnumerable<string> projects, CancellationToken token);
```
Comment (Matmalecki) [inline src/ReferenceAnalyzer.Core/IReferenceAnalyzer.cs:23]: This interface got really big fast. I think it should be split.

File: src/ReferenceAnalyzer.Core/ReferenceAnalyzer.cs [hunk 2]
```diff
@@ -33,10 +35,16 @@ public class ReferenceAnalyzer : IReferenceAnalyzer
         public bool ThrowOnCompilationFailures { get; set; } = true;
         public IProgress<double> ProgressReporter { get; set; } = new Progress<double>();
 
-        public async IAsyncEnumerable<ReferencesReport> AnalyzeAll(string solutionPath)
+        public IAsyncEnumerable<ReferencesReport> AnalyzeAll(string solutionPath)
```
Comment (Matmalecki) [inline src/ReferenceAnalyzer.Core/ReferenceAnalyzer.cs:38]: Why AnalyzeAll without CancellationToken is not async?

File: src/ReferenceAnalyzer.Core/ReferenceAnalyzer.cs [hunk 3]
```diff
@@ -103,31 +123,33 @@ private async Task<ReferencesReport> Analyze(Project project)
...
             var visitor = new ReferencesWalker(compilation, ignoreRules);
 
-            var treeAnalysis = new List<Task>(compilation.SyntaxTrees.Count());
-            var analysisTasks = compilation.SyntaxTrees
-                .Select(tree => Task.Run(async () => visitor.Visit(await tree.GetRootAsync())));
-            treeAnalysis.AddRange(analysisTasks);
-
-            await Task.WhenAll(treeAnalysis);
+            foreach (var tree in compilation.SyntaxTrees)
+                visitor.Visit(await tree.GetRootAsync(token));
```
Comment (Matmalecki) [inline src/ReferenceAnalyzer.Core/ReferenceAnalyzer.cs:127]: Add braces to foreach.

File: src/ReferenceAnalyzer.Core/ReferenceAnalyzer.cs [hunk 4]
```diff
@@ -181,22 +203,38 @@ private async Task<Compilation> GetCompilation(Project project, string outputPat
...
         {
+            if (projects == null)
+                throw new ArgumentNullException(nameof(projects));
+
+            token.Register(() => ProgressReporter.Report(0));
+
             var analyzedProjects = 0;
             var totalProjects = projects.Count();
-            ProgressReporter.Report(0);
+            ProgressReporter.Report(-1);
```
Comment (Matmalecki) [inline src/ReferenceAnalyzer.Core/ReferenceAnalyzer.cs:232]: Maybe it is time to add some const for Report() arguments. I had to think for a minute what does -1 mean in this context.

**Review by Matmalecki — COMMENTED** [review]
Most of it looks good. Also github actions left couple of warnings that could be solved right now.

File: src/ReferenceAnalyzer.Core/ReferenceAnalyzer.cs (hunk 2 above)
Comment (Artur-Laskowski) [reply src/ReferenceAnalyzer.Core/ReferenceAnalyzer.cs:39]: I'd have to wrap it's overload call in `await foreach` and `yield` to make it async method. Not worth it IMO.

**Pull request merged**